analyzers/src/SonarAnalyzer.Common/RegularExpressions/RegexContext.cs

```diff
@@ -0,0 +1,118 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Text.RegularExpressions;
+
+namespace SonarAnalyzer.RegularExpressions;
+
+internal sealed class RegexContext
+{
+    private static readonly RegexOptions TestMask = (RegexOptions)int.MinValue ^ RegexOptions.Compiled;
+
+    public SyntaxNode PatternNode { get; }
+    public string Pattern { get; }
+    public SyntaxNode OptionsNode { get; }
+    public RegexOptions? Options { get; }
+    public Regex Regex { get; }
+    public Exception ParseError { get; }
+
+    public RegexContext(
+        SyntaxNode patternNode,
+        string pattern,
+        SyntaxNode optionsNode,
+        RegexOptions? options)
+    {
+        PatternNode = patternNode;
+        Pattern = pattern;
+        OptionsNode = optionsNode;
+        Options = options;
+
+        if (!string.IsNullOrEmpty(pattern))
+        {
+            try
+            {
+                Regex = new(Pattern, options.GetValueOrDefault() & TestMask, TimeSpan.FromMilliseconds(100));
+            }
+            catch (Exception x)
+            {
+                ParseError = x;
+            }
+        }
+    }
+
+    public static RegexContext FromCtor<TSyntaxKind>(SyntaxNode node, SemanticModel model, ILanguageFacade<TSyntaxKind> language) where TSyntaxKind : struct =>
+        model.GetSymbolInfo(node).Symbol is IMethodSymbol method
+        && method.IsConstructor()
+        && method.ContainingType.Is(KnownType.System_Text_RegularExpressions_Regex)
+            ? FromSymbol(method, node, model, language)
+            : null;
+
+    public static RegexContext FromMethod<TSyntaxKind>(SyntaxNode node, SemanticModel model, ILanguageFacade<TSyntaxKind> language) where TSyntaxKind : struct =>
+        language.Syntax.NodeIdentifier(node).GetValueOrDefault().Text is { } name
+        && MatchMethods.Any(x => x.Equals(name, language.NameComparison))
+        && model.GetSymbolInfo(node).Symbol is IMethodSymbol { IsStatic: true } method
+        && method.ContainingType.Is(KnownType.System_Text_RegularExpressions_Regex)
+            ? FromSymbol(method, node, model, language)
+            : null;
+
+    public static RegexContext FromAttribute<TSyntaxKind>(SyntaxNode node, SemanticModel model, ILanguageFacade<TSyntaxKind> language) where TSyntaxKind : struct
+    {
+        if (model.GetSymbolInfo(node).Symbol is IMethodSymbol method
+            && method.IsInType(KnownType.System_ComponentModel_DataAnnotations_RegularExpressionAttribute))
+        {
+            var parameters = language.MethodParameterLookup(node, method);
+            var pattern = TryGetNonParamsSyntax(method, parameters, "pattern");
+            return new RegexContext(
+                pattern,
+                language.FindConstantValue(model, pattern) as string,
+                null,
+                null);
+        }
+        return null;
```

Comment:
Please add explicit `else { ... }` to be clear that each branch defines it's return value